src/category_theory/closed/functor.lean

```diff
@@ -0,0 +1,177 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+
+import category_theory.closed.cartesian
+import category_theory.limits.preserves.shapes.binary_products
+import category_theory.adjunction.fully_faithful
+
+/-!
+# Cartesian closed functors
+
+Define the exponential comparison morphisms for a functor which preserves binary products, and use
+them to define a cartesian closed functor: one which (naturally) preserves exponentials.
+
+Define the Frobenius morphism, and show it is an isomorphism iff the exponential comparison is an
+isomorphism.
+
+## TODO
+Some of the results here are true more generally for closed objects and for closed monoidal
+categories, and these could be generalised.
+
+## References
+https://ncatlab.org/nlab/show/Frobenius+reciprocity
+
+## Tags
+Frobenius reciprocity, cartesian closed functor
+
+-/
+
+namespace category_theory
+
+open category limits cartesian_closed
+
+universes v u u'
+variables {C : Type u} [category.{v} C]
+variables {D : Type u'} [category.{v} D]
+
+variables [has_finite_products C] [has_finite_products D]
+
+variables (F : C ⥤ D) {L : D ⥤ C}
+
+noncomputable theory
+
+/--
+The Frobenius morphism for an adjunction `L ⊣ F` at `A` is given by the morphism
+
+    L(FA ⨯ B) ⟶ LFA ⨯ LB ⟶ A ⨯ LB
+
+natural in `B`, where the first morphism is the product comparison and the latter uses the counit
+of the adjunction.
+
+We will show that if `C` and `D` are cartesian closed, then this morphism is an isomorphism for all
+`A` iff `F` is a cartesian closed functor, i.e. it preserves exponentials.
+-/
+def frobenius_morphism (h : L ⊣ F) (A : C) :
+  prod.functor.obj (F.obj A) ⋙ L ⟶ L ⋙ prod.functor.obj A :=
+prod_comparison_nat_trans L (F.obj A) ≫ whisker_left _ (prod.functor.map (h.counit.app _))
+
+/--
+If `F` is full and faithful and has a left adjoint `L` which preserves binary products
+-/
+instance frobenius_morphism_iso_of_preserves_finite_products (h : L ⊣ F) (A : C)
+  [preserves_limits_of_shape (discrete walking_pair) L] [full F] [faithful F] :
+is_iso (frobenius_morphism F h A) :=
+begin
+  apply nat_iso.is_iso_of_is_iso_app _,
+  intro B,
+  dsimp [frobenius_morphism],
+  apply_instance,
```

Comment:
By the way Johan, this is how I knew that the preserves products assumption was enough to make `is_iso prod_comparison_nat_trans`